man/RNAaliduplex.1.in
.TH "RNAaliduplex" "l" "0.0.1" "Ivo Hofacker" "Vienna RNA"
.SH "NAME"
.LP 
RNAaliduplex \- Predict conserved RNA-RNA interactions between two alignments
.SH "SYNTAX"
.LP 
RNAaliduplex    [\fI\-e range\fP] [\fIstandard_options\fP] <\fIfile1.aln\fP> <\fIfile2.aln\fP>

.SH "DESCRIPTION"
.LP 
\fIRNAaliduplex\fR reads two alignments of RNA sequences in CLUSTAL format and predicts optimal and suboptimal binding sites, hybridization energies and the corresponding structures. The calculation takes only inter\-molecular base pairs into account, for the general case use \fBRNAcofold\fR. The use of alignments allows to focus on binding sites that are evolutionary conserved. Note, that the two input alignments need to have equal number of sequences and the same order, i.e. the 1st sequence in file1 will be hybridized to the 1st in file2 etc.
.LP 
The computed binding sites, energies, and structures are written to stdout, one structure per line.
Each line consist of: The structure in dot bracket format with a "&" separating the two strands. The range of the structure in the two sequences in the format  "from,to : from,to"; the energy of duplex structure in kcal/mol.
.br 
The format is especially useful for computing the hybrid structure between a small probe sequence and a long target sequence.
.SH "OPTIONS"
.LP 
.TP 
\fB\-e\fR \fIrange\fR 
Compute suboptimal structures with energy with \fIrange\fR kcal/mol of the optimum.
Default is calculation of mfe structure only.
.PP 
For the standard options, such as 
\fB\-T\fR, \fB\-4\fR, \fB\-noGU\fR, \-\fBnoCloseGU\fR,
\fB\-P\fR see the \fBRNAfold\fR man page.

.SH "VERSION"
This man page documents version @VERSION@ Vienna RNA Package.
.SH "AUTHORS"
.LP 
Ivo Hofacker <ivo@tbi.univie.ac.at>
.SH "SEE ALSO"
.LP 
RNAduplex(l) RNAcofold(l) RNAalifold(l)


NEWS
Upcoming 
-----------
- Updated parameters for both RNA and DNA using the data compiled by 
  David Mathews <David_Mathews@urmc.rochester.edu>. 

Version 1.6
------------
- The RNAforester program for tree-alignments of RNA structures is now
  distributed with the Vienna RNA package, see the RNAforester
  subdirectory for more information. RNAforester was written by Matthias
  Hoechsmann <mhoechsm@techfak.uni-bielefeld.de> 
- The Kinfold program for stochastic simulation of folding trajectories is
  no included in the package, see the Kinfold subdirectory.
- cofolding of two structures now supports suboptimal folding and 
  partition function folding (still experimental). 
  ATTENTION: Energies of hybrid structures now inlcude the Duplex-initiation
  energy, which was neglected in previous version.
- RNAplfold is a partition function variant of RNALfold. It computes the
  mean probability of a (local) base pair averaged over all sequence
  windows that contain the pair.
- new utilities to color alignments and consensus structures  

Version 1.5pre 
-----------
- ViennaRNA now uses autoconfig generated configure scripts for even better
  portability (should compile on any UNIX, Linux, MacOS X, Windows with
  Cygwin).
- The new RNAalifold program predicts consensus structures for a set of
  aligned sequences.
- Complete suboptimal folding is now integrated in the library.
- Beginning support for co-folding of two strands: energy_of_struct() and
  RNAeval can now compute energies of duplex structures.
- RNAcofold predicts hybrid structures of two RNA strands
- RNAduplex predicts hybrid structures, while allowing only inter-molecular
  base pairs (useful for finding potential binding sites)
- RNALfold predicts locally stable structures in long sequences.
- Major changes to Perl module. See the pod documentation (perldoc RNA).
- RNAsubopt can do stochastic backtracking to produce samples of suboptimal
  structures with Boltzmann statistics.
- New utilities to rotate secondary structure plots and annotate them with
  reliability data.
- Various small bug fixes

Version 1.4
-----------
- New Turner parameters as described in Mathews et.al. JMB v288, 1999.
  Small changes to format of parameter files (old param files won't work!)
- mfe and suboptimal folding will produce only structures without isolated
  pairs if noLonelyPairs=1 (-noLP option), for partition function folding
  pairs that can only occur as isolated pairs are not formed.
- setting dangles=3 (-d3 option) will allow co-axial stacking of adjacent
  helices in mfe folding and energy_of_struct().

Version 1.3.1
-------------
- RNAheat would produce spikes in the specific heat because dangling end
  energies did not go smoothly to 0. 
- PS dot plots now have an option to use a log scale (edit _dp.ps file
  and set logscale to true).

Version 1.3
-----------
- Secondary structure plots now use E. Bruccoleri's naview routines for
  layout by default. New utility RNAplot produces secondary structure plots 
  from structures in bracket notation with several options.
- New -d2 option in RNAfold and RNAeval sets dangles=2, which makes 
  energy_of_struct() and fold() treat dangling ends as in pf_fold().
  -noLP option in RNAfold etc sets noLonelyPairs=1, which avoids most
  structures containing lonely base pairs (helices of length 1).
- new utility functions pack_structure() unpack_structure() make_pair_table()
  and bp_distance().  RNAdistance adds bp_distance() via -DP switch.
- First release of RNAsubopt for complete suboptimal folding.
- fixed bug in asymmetry penalty for interior loops. 
- Default compilation now uses doubles for partition function folding.

Version 1.2.1
-------------
- Fixed bug in version 1.2 of the RNAheat program causing overflow errors
  for most input sequences. 
- The PS_dot_plot() and PS_rna_plot() routines now return an int. The return 
  value is 0 if the file could not be written, 1 otherwise.
- This version contains the alpha version of a perl5 module, which let's you
  access all the capabilities of the Vienna RNA library from perl scripts.

Version 1.2
-----------
- New energy parameters from (Walter et.al 1994). 
- Energy parameters can be read from file.
- RNAeval and energy_of_struct() support logarithmic energy function for 
  multi-loops.
- gmlRNA() produces secondary structure drawing in gml (Graph Meta Language). 
- Many bug fixes.


Progs/RNAaliduplex.c
/* Last changed Time-stamp: <2005-07-23 16:50:24 ivo> */
/*                
	     Compute duplex structure of two RNA strands

			   c Ivo L Hofacker
			  Vienna RNA package
*/

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <unistd.h>
#include <string.h>
#include "duplex.h"
#include "fold.h"
#include "fold_vars.h"
#include "utils.h"
#include "aln_util.h"

extern void  read_parameter_file(const char fname[]);
extern int subopt_sorted;
extern duplexT aliduplexfold(const char *s1[], const char *s2[]);
static void  print_struc(duplexT const *dup);
/*@unused@*/
static char rcsid[] = "$Id: RNAaliduplex.c,v 1.1 2007/08/26 10:08:44 ivo Exp $";

static char  scale[] = "....,....1....,....2....,....3....,....4"
                       "....,....5....,....6....,....7....,....8";

static void usage(void);

/*--------------------------------------------------------------------------*/
#define MAX_NUM_NAMES    500

int main(int argc, char *argv[])
{
  char *AS1[MAX_NUM_NAMES], *AS2[MAX_NUM_NAMES], 
    *names1[MAX_NUM_NAMES], *names2[MAX_NUM_NAMES];
  char  fname[13];
  char  *ParamFile=NULL;
  int   i, r, n_seq, n_seq2;
  double deltaf;
  double kT, sfact=1.07;
  int   pf=0, istty, delta=-1;
  int noconv=0;
  duplexT mfe, *subopt;
  FILE  *file1=NULL, *file2=NULL; /* input alignments */

  for (i=1; i<argc; i++) {
    if (argv[i][0]=='-') 
      switch ( argv[i][1] )
	{
	case 'T':  if (argv[i][2]!='\0') usage();
	  if(i==argc-1) usage();
	  r=sscanf(argv[++i], "%lf", &temperature);
	  if (!r) usage();
	  break;
	case 'p':
	  fprintf(stderr, "partition function folding not yet implemented\n");
	  usage();
	  pf=1;
	  if (argv[i][2]!='\0')
	    (void) sscanf(argv[i]+2, "%d", &do_backtrack);
	  break;
	case 'n':
	  if ( strcmp(argv[i], "-noGU")==0) noGU=1;
	  if ( strcmp(argv[i], "-noCloseGU")==0) no_closingGU=1;
	  if ( strcmp(argv[i], "-noLP")==0) noLonelyPairs=1;
	  if ( strcmp(argv[i], "-noconv")==0) noconv=1;
	  break;
	case '4':
	  tetra_loop=0;
	  break;
	case 'e':
	  if (i>=argc-1) usage();
	  r=sscanf(argv[++i], "%lf", &deltaf);
	  if (r!=1) usage();
	  delta = (int) (0.1+deltaf*100);
	  break;
	case 's': subopt_sorted=1;
	  break;
	case 'd': dangles=0;
	  if (argv[i][2]!='\0') {
	    r=sscanf(argv[i]+2, "%d", &dangles);
	    if (r!=1) usage();
	  }
	  break;
	case 'P':
	  if (i==argc-1) usage();
	  ParamFile = argv[++i];
	  break;
	default: usage();
	} 
    else { /* doesn't start with '-' should be filename */
      if (i>argc-1) usage();
      file1 = fopen(argv[i], "r");
      if (file1 == NULL) {
        fprintf(stderr, "can't open %s\n", argv[i]);
        usage();
      }
      file2 = fopen(argv[++i], "r");
      if (file1 == NULL) {
        fprintf(stderr, "can't open %s\n", argv[i]);
        usage();
      }
    }
  }

  if (!(file1 && file2)) {
    fprintf(stderr, "No input files");
    usage();
  }
  if (ParamFile != NULL)
    read_parameter_file(ParamFile);
   
  istty = isatty(fileno(stdout))&&isatty(fileno(stdin));
	
  n_seq = read_clustal(file1, AS1, names1);
  n_seq2 = read_clustal(file2, AS2, names2);
  fclose(file1); fclose(file2);

  if (n_seq != n_seq2) nrerror("unequal number of seqs in alignments");

  update_fold_params();

  if (delta>=0) {
    duplexT *sub;
    subopt = aliduplex_subopt(AS1, AS2, delta, 5);
    for (sub=subopt; sub->i >0; sub++) {
      print_struc(sub);
      free(sub->structure);
    }
    free(subopt);
  } else {
    mfe = aliduplexfold(AS1, AS2);
    print_struc(&mfe);
    free(mfe.structure);
  }
  for (i=0; AS1[i]; i++) {
    free(AS1[i]); free(names1[i]);
    free(AS2[i]); free(names2[i]);
  }
  return 0;
}

static void print_struc(duplexT const *dup) {
  int l1;
  l1 = strchr(dup->structure, '&')-dup->structure;
  printf("%s %3d,%-3d : %3d,%-3d (%5.2f)\n", dup->structure, dup->i+1-l1,
	 dup->i, dup->j, dup->j+strlen(dup->structure)-l1-2, dup->energy);
}
    
static void usage(void)
{
  nrerror("usage:\n"
	  "RNAaliduplex [-e range] [-s] [standard options] file1.aln file2.aln\n");
}
